### PR TITLE
Trace recording and multi-stepping for interpreter

### PR DESCRIPTION
**Pull request by dcattaruzza**
Added functionality for multi-stepping (see h command when running interpreter) and printing traces (goto-trace and json)
fixes for goto-trace and json_goto_trace when obtaining data from goto-programs

### CODE REVIEW

File: src/goto-programs/goto_trace.cpp [hunk 1]
```diff
@@ -75,7 +75,11 @@ void goto_trace_stept::output(
   case goto_trace_stept::SHARED_WRITE: out << "SHARED WRITE"; break;
   case goto_trace_stept::FUNCTION_CALL: out << "FUNCTION CALL"; break;
   case goto_trace_stept::FUNCTION_RETURN: out << "FUNCTION RETURN"; break;
-  default: assert(false);
+  case goto_trace_stept::DEAD: out << "DEAD" << std::endl; return;
+  default: {
```
Comment (tautschnig) [inline src/goto-programs/goto_trace.cpp]: My suggestion would be to omit those default cases - the compiler will warn about missing cases.

File: src/goto-programs/interpreter.cpp [hunk 2]
```diff
@@ -49,11 +53,31 @@ void interpretert::operator()()
     
   done=false;
   
-  while(!done)
+  try {
+	std::cout << "Type h for help";
```
Comment (tautschnig) [inline src/goto-programs/interpreter.cpp]: nit pick: indentation seems wrong

File: src/goto-programs/interpreter.cpp [hunk 3]
```diff
@@ -82,7 +106,6 @@ void interpretert::show_state()
   }
   else
     function->second.body.output_instruction(ns, function->first, std::cout, PC);
-
```
Comment (tautschnig) [inline src/goto-programs/interpreter.cpp]: Whitespace change?

File: src/goto-programs/interpreter.cpp [hunk 4]
```diff
@@ -112,6 +135,42 @@ void interpretert::command()
 
   if(ch=='q')
     done=true;
+  else if (ch=='h') {
```
Comment (tautschnig) [inline src/goto-programs/interpreter.cpp]: Nit picking: put opening { on new line

File: src/goto-programs/interpreter.cpp [hunk 5]
```diff
@@ -189,28 +260,34 @@ void interpretert::step()
...
     throw "ATOMIC_BEGIN not yet implemented";
     
   case ATOMIC_END:
+    trace_step.type=goto_trace_stept::ATOMIC_END;
     throw "ATOMIC_END not yet implemented";
     
   case DEAD:
-    throw "DEAD not yet implemented";
+    trace_step.type=goto_trace_stept::DEAD;
+    break;//throw "DEAD not yet implemented";
```
Comment (tautschnig) [inline src/goto-programs/interpreter.cpp]: What does the remaining comment tell us? Is it implemented or not?

File: src/goto-programs/interpreter.cpp [hunk 6]
```diff
@@ -317,8 +394,41 @@ void interpretert::execute_assign()
...
+          lhs_at_pc=true_exprt();
+        else 
+          lhs_at_pc=false_exprt();
+      }
+      else
+      {
+          lhs_at_pc=from_integer(rhs[0],size_type());
+      }
+      trace_step.full_lhs_value=lhs_at_pc;
+      trace_step.lhs_object_value = lhs_at_pc;
```
Comment (tautschnig) [inline src/goto-programs/interpreter.cpp]: Nit pick: no whitespace around =

File: src/goto-programs/interpreter_class.h [hunk 7]
```diff
@@ -81,13 +83,18 @@ class interpretert
     memory_mapt local_map;
     unsigned old_stack_pointer;
   };
-
```
Comment (tautschnig) [inline src/goto-programs/interpreter_class.h]: Whitespace change?

File: src/goto-programs/interpreter_class.h [hunk 8]
```diff
@@ -81,13 +83,18 @@ class interpretert
     memory_mapt local_map;
     unsigned old_stack_pointer;
   };
-  
+
   typedef std::stack<stack_framet> call_stackt;
-  call_stackt call_stack;
   
+  call_stackt call_stack;
```
Comment (tautschnig) [inline src/goto-programs/interpreter_class.h]: Line moved around blank line?

File: src/goto-programs/json_goto_trace.cpp [hunk 9]
```diff
@@ -27,7 +27,8 @@ Function: json
  Purpose:
 
 \*******************************************************************/
-
```
Comment (tautschnig) [inline src/goto-programs/json_goto_trace.cpp]: blank lines are not forbidden ...

**Comment by peterschrammel** [comment]
@dcattaruzza, is this subsumed by the larger interpreter patch?

**Comment by dcattaruzza** [comment]
This functionality was moved to the test-gen-support branch. I'm currently working on some additional fixes for missing functionality/formatting. Features that should go back to master will then be included in a new PR

**Pull request closed without merging**